### PR TITLE
adapter: export audit events to Segment

### PR DESCRIPTION
**Pull request by benesch**
Details in commit messages.

Fixes https://github.com/MaterializeInc/materialize/issues/11371.

### Motivation

  * This PR fixes a recognized bug.

### Tips for reviewer

@mjibson I figured you could scrutinize the code, and @ahelium you could validate that these events will be sufficient.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Review by ahelium — APPROVED** [review]
Thank you so much @benesch! This is incredible. 

To make sure I'm following, these audit events:
```
materialize=> select * from mz_audit_events; 
 id | event_type |    object_type    |                                     event_details                                      |         user         |        occurred_at         
----+------------+-------------------+----------------------------------------------------------------------------------------+----------------------+----------------------------
 11 | create     | cluster           | {"name":"analytics"}                                                                   | anna@materialize.com | 2022-09-26 13:42:08.475+00
 10 | create     | view              | {"database":"materialize","item":"t","schema":"public"}                                | anna@materialize.com | 2022-09-01 22:05:09.438+00
```

Will be translated into the following track calls:
```
{ 
   .... 
   "event": "Cluster Created", 
   ....
   "properties": {
     "organization_id": "b6af3eaf-d3c3-48f9-805c-a18b19d0ab95",
     "event_source":"environmentd",
     "details":{"name":"analytics"}
   },
   ....
   "type": "track",
   "userId": "d15dc710-0d47-4201-b87d-a16ed04acf91",
   ....
}


{
   ....
   "event": "View Created",
   ....
   "properties": {
     "organization_id": "b6af3eaf-d3c3-48f9-805c-a18b19d0ab95",
     "event_source":"environmentd",
     "details": {"database":"materialize","item":"t","schema":"public"}
   },
   ....
   "type": "track",
   "userId": "d15dc710-0d47-4201-b87d-a16ed04acf91",
   ....
}

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -3276,18 +3286,23 @@ impl<S: Append> Catalog<S> {
...
         let mut builtin_table_updates = vec![];
+        let mut audit_events = vec![];
         let mut actions = Vec::with_capacity(ops.len());
         let mut storage = self.storage().await;
         let mut tx = storage.transaction().await?;
 
         fn sql_type_to_object_type(sql_type: SqlCatalogItemType) -> ObjectType {
             match sql_type {
-                SqlCatalogItemType::View => ObjectType::View,
+                SqlCatalogItemType::Connection => ObjectType::Connection,
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/catalog.rs:3296]: (Surprisingly) those aren't SqlCatalogItemTypes, so we aren't missing them. They get tracked elsewhere.

**Review by mjibson — APPROVED** [review]
Why does segment need the uuid of the user instead of the human-readable string name? Should we be logging that alongside the string username in the audit log too?

**Comment by benesch** [comment]
> Why does segment need the uuid of the user instead of the human-readable string name? Should we be logging that alongside the string username in the audit log too?

That's just what Segment prefers! You associate track events with a stable UUID. Separately you can update the mapping from UUID -> email/name/other attributes—which we do over in the cloud repo.

**Comment by benesch** [comment]
That's exactly right, @ahelium! I just tested this build in cloud production and verified that it works as promised. I'm going to merge as is, but feel free to raise issues/open PRs to further adjust the event format.

**Pull request merged**